Automattic/sensei
# Add locked lesson free course notices

Part of #4445

### Changes proposed in this Pull Request

* It introduces notices  (logged out and not enrolled students) in the lesson pages for free courses.
* It also contains some redirects logic if the user is logged out.
  * If they click on "Sign in", they log in, then they will be redirected to the same page they were.
  * If they click on "Take course", they log in, then they will be redirected to the course page, with the button "Start course" instead of "Take course" (or any other name they used).
* Notice that the redirects after login may not work if users are using a custom login/registration page. It's was implemented using `redirect_to` query string, which is the same used by WP core login.
* It also refactors some code to use the new `sensei_user_registration_url` function.

### Testing instructions


* Enable the feature flag: `add_filter( 'sensei_feature_flag_course_theme', '__return_true' );`
* Create a course.
* Add at least one lesson.
* Select the "Sensei Theme" as the "Course Theme" (sidebar).
* Logged out, access a lesson page, and make sure you see a notice with "Take course" and "Sign in" buttons.
* Using the Sign in button, make sure you're redirected back to the lesson page you were.
* Using the Take course button, make sure you're redirected to the course with a button "Start course" instead of "Take course" (just the label was replaced for that redirect case).
* Now, logged in as a student, try to access a lesson page, and make sure you see a notice with the "Take course" button only.
* Make sure that by clicking on that button, you get enrolled in the course. When https://github.com/Automattic/sensei/pull/4496 is completed, the user will be redirected to the first lesson after the enrollment.


### New/Updated Hooks

* `sensei_login_success_redirect_url` - The old `sesei_login_success_redirect_url` renamed.


### Deprecated Code

* `sesei_login_success_redirect_url` - Deprecated because the typo in the filter name.


### Screenshot / Video

https://user-images.githubusercontent.com/876340/145272704-40b977d8-71e1-4011-82f9-6439125be868.mov

On 0:13, notice that it will redirect to the first lesson after https://github.com/Automattic/sensei/pull/4496 is completed.

#### Another case, in a course with course prerequisite.

<img width="940" alt="Screen Shot 2021-12-08 at 16 27 08" src="https://user-images.githubusercontent.com/876340/145272783-6159274e-9189-4823-b311-906819158e6b.png">
